File: src/lib/excelExport.ts
import ExcelJS from "exceljs";
import { ComparisonResult } from "./types";

const COLORS = {
  added: { bg: "C6EFCE", font: "006100", label: "ADDED" },
  deleted: { bg: "FFC7CE", font: "9C0006", label: "DELETED" },
  modified: { bg: "FFE0B2", font: "E65100", label: "MODIFIED" },
  unchanged: { bg: "E0E0E0", font: "424242", label: "NO CHANGE" },
  headerBg: "003366",
  headerFont: "FFFFFF",
};

function stringify(value: unknown): string {
  if (value === null || value === undefined) return "";
  return String(value);
}

export async function exportComparisonToExcel(
  result: ComparisonResult
): Promise<Blob> {
  const workbook = new ExcelJS.Workbook();

  // ── Sheet 1: Summary Dashboard ──
  const summarySheet = workbook.addWorksheet("Summary", {
    properties: { tabColor: { argb: "003366" } },
  });

  summarySheet.columns = [
    { header: "", key: "label", width: 25 },
    { header: "", key: "value", width: 30 },
  ];

  const titleRow = summarySheet.addRow(["FILE COMPARISON REPORT", ""]);
  titleRow.font = { bold: true, size: 16, color: { argb: "003366" } };
  summarySheet.mergeCells("A1:B1");

  summarySheet.addRow([]);
  summarySheet.addRow(["Before File:", result.beforeFileName]);
  summarySheet.addRow(["After File:", result.afterFileName]);
  summarySheet.addRow(["Compared At:", new Date(result.comparedAt).toLocaleString()]);
  summarySheet.addRow(["Key Column:", result.keyColumn]);
  summarySheet.addRow([]);

  const statsHeader = summarySheet.addRow(["CHANGE SUMMARY", ""]);
  statsHeader.font = { bold: true, size: 13, color: { argb: "003366" } };
  summarySheet.mergeCells(`A${statsHeader.number}:B${statsHeader.number}`);

  summarySheet.addRow([]);

  const statsRows = [
    ["Total Records", result.summary.total],
    ["Added (New)", result.summary.added],
    ["Deleted (Removed)", result.summary.deleted],
    ["Modified (Changed)", result.summary.modified],
    ["Unchanged", result.summary.unchanged],
  ];

  const statColors: Record<string, string> = {
    "Added (New)": COLORS.added.bg,
    "Deleted (Removed)": COLORS.deleted.bg,
    "Modified (Changed)": COLORS.modified.bg,
    Unchanged: COLORS.unchanged.bg,
  };

  for (const [label, value] of statsRows) {
    const row = summarySheet.addRow([label, value]);
    row.font = { bold: true, size: 11 };
    const colorKey = label as string;
    if (statColors[colorKey]) {
      row.getCell(1).fill = {
        type: "pattern",
        pattern: "solid",
        fgColor: { argb: statColors[colorKey] },
      };
      row.getCell(2).fill = {
        type: "pattern",
        pattern: "solid",
        fgColor: { argb: statColors[colorKey] },
      };
    }
  }

  // ── Sheet 2: Detailed Comparison ──
  const detailSheet = workbook.addWorksheet("Comparison Details", {
    properties: { tabColor: { argb: "0066CC" } },
  });

  const detailHeaders = ["Status", "Key", ...result.columns];
  const headerRow = detailSheet.addRow(detailHeaders);

  headerRow.eachCell((cell) => {
    cell.fill = {
      type: "pattern",
      pattern: "solid",
      fgColor: { argb: COLORS.headerBg },
    };
    cell.font = { bold: true, color: { argb: COLORS.headerFont }, size: 11 };
    cell.alignment = { horizontal: "center", vertical: "middle" };
    cell.border = {
      bottom: { style: "thin", color: { argb: "000000" } },
    };
  });

  // Set column widths
  detailSheet.columns = detailHeaders.map((h) => ({
    header: h,
    key: h,
    width: Math.max(15, h.length + 5),
  }));

  for (const compRow of result.rows) {
    const color = COLORS[compRow.status];
    const data = compRow.afterData || compRow.beforeData || {};
    const rowValues: string[] = [
      color.label,
      compRow.keyValue,
      ...result.columns.map((col) => stringify(data[col])),
    ];

    const excelRow = detailSheet.addRow(rowValues);

    excelRow.eachCell((cell) => {
      cell.fill = {
        type: "pattern",
        pattern: "solid",
        fgColor: { argb: color.bg },
      };
      cell.font = { color: { argb: color.font } };
      cell.border = {
        bottom: { style: "hair", color: { argb: "CCCCCC" } },
      };
    });

    // Highlight modified cells specifically
    if (compRow.status === "modified") {
      for (const change of compRow.changes) {
        const colIdx = result.columns.indexOf(change.column);
        if (colIdx >= 0) {
          const cell = excelRow.getCell(colIdx + 3); // +3 for Status and Key columns
          cell.value = `${stringify(change.oldValue)} → ${stringify(change.newValue)}`;
          cell.font = { bold: true, color: { argb: "E65100" } };
          cell.fill = {
            type: "pattern",
            pattern: "solid",
            fgColor: { argb: "FFD180" },
          };
        }
      }
    }
  }

  // Auto-filter
  detailSheet.autoFilter = {
    from: { row: 1, column: 1 },
    to: { row: result.rows.length + 1, column: detailHeaders.length },
  };

  // Freeze the header row
  detailSheet.views = [{ state: "frozen", ySplit: 1, xSplit: 2 }];

  // ── Sheet 3: Changes Only ──
  const changesSheet = workbook.addWorksheet("Changes Only", {
    properties: { tabColor: { argb: "F58220" } },
  });

  const changesHeaders = ["Status", "Key", "Column", "Before Value", "After Value"];
  const changesHeaderRow = changesSheet.addRow(changesHeaders);

  changesHeaderRow.eachCell((cell) => {
    cell.fill = {
      type: "pattern",
      pattern: "solid",
      fgColor: { argb: COLORS.headerBg },
    };
    cell.font = { bold: true, color: { argb: COLORS.headerFont }, size: 11 };
    cell.alignment = { horizontal: "center", vertical: "middle" };
  });

  changesSheet.columns = [
    { header: "Status", key: "Status", width: 15 },
    { header: "Key", key: "Key", width: 25 },
    { header: "Column", key: "Column", width: 25 },
    { header: "Before Value", key: "Before Value", width: 30 },
    { header: "After Value", key: "After Value", width: 30 },
  ];

  for (const compRow of result.rows) {
    if (compRow.status === "unchanged") continue;

    if (compRow.status === "modified") {
      for (const change of compRow.changes) {
        const row = changesSheet.addRow([
          COLORS.modified.label,
          compRow.keyValue,
          change.column,
          stringify(change.oldValue),
          stringify(change.newValue),
        ]);
        row.eachCell((cell) => {
          cell.fill = {
            type: "pattern",
            pattern: "solid",
            fgColor: { argb: COLORS.modified.bg },
          };
          cell.font = { color: { argb: COLORS.modified.font } };
        });
      }
    } else {
      const color = COLORS[compRow.status];
      const row = changesSheet.addRow([
        color.label,
        compRow.keyValue,
        "—",
        compRow.status === "deleted" ? "(entire row)" : "",
        compRow.status === "added" ? "(entire row)" : "",
      ]);
      row.eachCell((cell) => {
        cell.fill = {
          type: "pattern",
          pattern: "solid",
          fgColor: { argb: color.bg },
        };
        cell.font = { color: { argb: color.font } };
      });
    }
  }

  changesSheet.views = [{ state: "frozen", ySplit: 1 }];

  // Generate the file
  const arrayBuffer = await workbook.xlsx.writeBuffer();
  return new Blob([arrayBuffer], {
    type: "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  });
}

File: excel_export.py
"""Color-coded Excel export module using openpyxl."""

import io
from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from openpyxl.utils import get_column_letter


# Color configuration
COLORS = {
    "added":     {"bg": "C6EFCE", "font": "006100", "label": "ADDED"},
    "deleted":   {"bg": "FFC7CE", "font": "9C0006", "label": "DELETED"},
    "modified":  {"bg": "FFE0B2", "font": "E65100", "label": "MODIFIED"},
    "unchanged": {"bg": "E0E0E0", "font": "424242", "label": "NO CHANGE"},
}
HEADER_BG = "003366"
HEADER_FONT = "FFFFFF"


def _stringify(value):
    if value is None:
        return ""
    s = str(value)
    return "" if s == "nan" else s


def export_comparison(result, before_name, after_name):
    """
    Export comparison result to a color-coded Excel workbook.

    Returns an io.BytesIO buffer containing the .xlsx file.
    """
    wb = Workbook()

    # ── Sheet 1: Summary ──
    ws_summary = wb.active
    ws_summary.title = "Summary"
    ws_summary.sheet_properties.tabColor = "003366"

    ws_summary.column_dimensions["A"].width = 28
    ws_summary.column_dimensions["B"].width = 35

    title_font = Font(bold=True, size=16, color="003366")
    ws_summary.merge_cells("A1:B1")
    ws_summary["A1"].value = "FILE COMPARISON REPORT"
    ws_summary["A1"].font = title_font

    summary = result["summary"]
    info_rows = [
        ("Before File:", before_name),
        ("After File:", after_name),
        ("Compared At:", result["comparedAt"]),
        ("Key Column:", result["keyColumn"]),
    ]
    for i, (label, value) in enumerate(info_rows, start=3):
        ws_summary.cell(row=i, column=1, value=label).font = Font(bold=True, size=11)
        ws_summary.cell(row=i, column=2, value=value)

    # Stats header
    r = 8
    ws_summary.merge_cells(f"A{r}:B{r}")
    ws_summary.cell(row=r, column=1, value="CHANGE SUMMARY").font = Font(bold=True, size=13, color="003366")

    stat_rows = [
        ("Total Records",       summary["total"],     None),
        ("Added (New)",         summary["added"],     COLORS["added"]["bg"]),
        ("Deleted (Removed)",   summary["deleted"],   COLORS["deleted"]["bg"]),
        ("Modified (Changed)",  summary["modified"],  COLORS["modified"]["bg"]),
        ("Unchanged",           summary["unchanged"], COLORS["unchanged"]["bg"]),
    ]
    for i, (label, value, bg) in enumerate(stat_rows, start=10):
        c1 = ws_summary.cell(row=i, column=1, value=label)
        c2 = ws_summary.cell(row=i, column=2, value=value)
        c1.font = Font(bold=True, size=11)
        c2.font = Font(bold=True, size=11)
        if bg:
            fill = PatternFill(start_color=bg, end_color=bg, fill_type="solid")
            c1.fill = fill
            c2.fill = fill

    # ── Sheet 2: Comparison Details ──
    ws_detail = wb.create_sheet("Comparison Details")
    ws_detail.sheet_properties.tabColor = "0066CC"

    columns = result["columns"]
    headers = ["Status", "Key"] + columns
    header_fill = PatternFill(start_color=HEADER_BG, end_color=HEADER_BG, fill_type="solid")
    header_font = Font(bold=True, color=HEADER_FONT, size=11)
    thin_border = Border(bottom=Side(style="thin", color="000000"))

    for col_idx, header in enumerate(headers, start=1):
        cell = ws_detail.cell(row=1, column=col_idx, value=header)
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(horizontal="center", vertical="center")
        cell.border = thin_border
        ws_detail.column_dimensions[get_column_letter(col_idx)].width = max(15, len(header) + 5)

    hair_border = Border(bottom=Side(style="hair", color="CCCCCC"))

    for row_idx, comp_row in enumerate(result["rows"], start=2):
        status = comp_row["status"]
        color = COLORS[status]
        row_fill = PatternFill(start_color=color["bg"], end_color=color["bg"], fill_type="solid")
        row_font = Font(color=color["font"])

        data = comp_row["afterData"] or comp_row["beforeData"] or {}

        # Status cell
        cell = ws_detail.cell(row=row_idx, column=1, value=color["label"])
        cell.fill = row_fill
        cell.font = Font(bold=True, color=color["font"])
        cell.border = hair_border

        # Key cell
        cell = ws_detail.cell(row=row_idx, column=2, value=comp_row["keyValue"])
        cell.fill = row_fill
        cell.font = row_font
        cell.border = hair_border

        # Data cells
        changed_cols = {c["column"] for c in comp_row["changes"]}
        for col_idx, col_name in enumerate(columns, start=3):
            cell = ws_detail.cell(row=row_idx, column=col_idx, value=_stringify(data.get(col_name, "")))
            cell.fill = row_fill
            cell.font = row_font
            cell.border = hair_border

            # Highlight changed cells in modified rows
            if status == "modified" and col_name in changed_cols:
                change = next(c for c in comp_row["changes"] if c["column"] == col_name)
                cell.value = f"{_stringify(change['oldValue'])} → {_stringify(change['newValue'])}"
                cell.font = Font(bold=True, color="E65100")
                cell.fill = PatternFill(start_color="FFD180", end_color="FFD180", fill_type="solid")

    # Auto-filter and freeze
    ws_detail.auto_filter.ref = f"A1:{get_column_letter(len(headers))}{len(result['rows']) + 1}"
    ws_detail.freeze_panes = "C2"

    # ── Sheet 3: Changes Only ──
    ws_changes = wb.create_sheet("Changes Only")
    ws_changes.sheet_properties.tabColor = "F58220"

    change_headers = ["Status", "Key", "Column", "Before Value", "After Value"]
    for col_idx, header in enumerate(change_headers, start=1):
        cell = ws_changes.cell(row=1, column=col_idx, value=header)
        cell.fill = header_fill
        cell.font = header_font
        cell.alignment = Alignment(horizontal="center", vertical="center")

    ws_changes.column_dimensions["A"].width = 15
    ws_changes.column_dimensions["B"].width = 25
    ws_changes.column_dimensions["C"].width = 25
    ws_changes.column_dimensions["D"].width = 30
    ws_changes.column_dimensions["E"].width = 30

    change_row = 2
    for comp_row in result["rows"]:
        status = comp_row["status"]
        if status == "unchanged":
            continue

        color = COLORS[status]
        row_fill = PatternFill(start_color=color["bg"], end_color=color["bg"], fill_type="solid")
        row_font = Font(color=color["font"])

        if status == "modified":
            for change in comp_row["changes"]:
                for ci, val in enumerate([
                    color["label"],
                    comp_row["keyValue"],
                    change["column"],
                    _stringify(change["oldValue"]),
                    _stringify(change["newValue"]),
                ], start=1):
                    cell = ws_changes.cell(row=change_row, column=ci, value=val)
                    cell.fill = row_fill
                    cell.font = row_font
                change_row += 1
        else:
            values = [
                color["label"],
                comp_row["keyValue"],
                "—",
                "(entire row)" if status == "deleted" else "",
                "(entire row)" if status == "added" else "",
            ]
            for ci, val in enumerate(values, start=1):
                cell = ws_changes.cell(row=change_row, column=ci, value=val)
                cell.fill = row_fill
                cell.font = row_font
            change_row += 1

    ws_changes.freeze_panes = "A2"

    # Write to buffer
    buffer = io.BytesIO()
    wb.save(buffer)
    buffer.seek(0)
    return buffer
